### PR TITLE
Optimize TopLevelMethodDefs

### PR DESCRIPTION
**Pull request by gzm0**

### CODE REVIEW

**Comment by gzm0** [comment]
Diff on the test suite: https://gist.github.com/gzm0/2b538d39492352a223e8047e75226600

Notable: We're now able to eliminate the scala.scalajs.runtime package object entirely.

**Comment by gzm0** [comment]
I am a bit concerned about the spurious diffs (names & ordering) :-/

**Comment by sjrd** [comment]
From what I can tell from a mobile phone, this looks good. I'll have a better look once I have a computer, the day after tomorrow.

**Comment by gzm0** [comment]
Marking as draft, likely affected by a similar issue than #4774.

**Comment by gzm0** [comment]
Not affected by #4774, but I still added a test case for good measure.

File: linker/shared/src/test/scala/org/scalajs/linker/IncrementalTest.scala
Comment (sjrd) [inline linker/shared/src/test/scala/org/scalajs/linker/IncrementalTest.scala:432]: Same formatting comments here as in the other PR.

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala:798]: Consider using `newMethod.pos` here, so that you don't have to keep track of positions in `methods`.

Or else store the `TopLevelMethodExportDef` as values of `methods` instead.

**Review by sjrd — CHANGES_REQUESTED** [review]

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala
Comment (gzm0) [reply linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala:798]: > Consider using newMethod.pos here,

Wouldn't that move the position form the annotation (that triggered the export) to the method definition?

> Or else store the TopLevelMethodExportDef as values of methods instead.

I'm not sure I understand. You mean as second element of the tuple (where pos is right now)? Or make a separate JSTopLevelMethodImpl (that's not so easy, because `Processable` requires a `version` member).

Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala:798]: Ah yes indeed. Both points make sense.

Let's leave it as is, then.

**Review by sjrd — APPROVED** [review]
LGTM, but it will need a rebase once #4775 is merged.

**Comment by sjrd** [comment]
This can be rebased now.

**Comment by gzm0** [comment]
Rebased.

**Review by sjrd — APPROVED** [review]

**Comment by sjrd** [comment]
There seem to be a number of non-spurious CI failures, with the following exception (testSuite in `CommonJSModule` mode):
```
[info] Fast optimizing .../test-suite/js/.2.12/target/scala-2.12/scalajs-test-suite-test-fastopt
[error] java.util.NoSuchElementException: key not found: ClassName<org.scalajs.testsuite.jsinterop.ExportedDefaultArgClass$>
[error] 	at scala.collection.MapLike.default(MapLike.scala:236)
[error] 	at scala.collection.MapLike.default$(MapLike.scala:235)
[error] 	at scala.collection.AbstractMap.default(Map.scala:65)
[error] 	at scala.collection.MapLike.apply(MapLike.scala:144)
[error] 	at scala.collection.MapLike.apply$(MapLike.scala:143)
[error] 	at scala.collection.AbstractMap.apply(Map.scala:65)
[error] 	at org.scalajs.linker.frontend.modulesplitter.Tagger.tag(Tagger.scala:93)
[error] 	at org.scalajs.linker.frontend.modulesplitter.Tagger.staticEdge(Tagger.scala:109)
[error] 	at org.scalajs.linker.frontend.modulesplitter.Tagger.$anonfun$tagEntryPoints$3(Tagger.scala:120)
[error] 	at org.scalajs.linker.frontend.modulesplitter.Tagger.$anonfun$tagEntryPoints$3$adapted(Tagger.scala:118)
[error] 	at scala.collection.immutable.HashSet$HashSet1.foreach(HashSet.scala:335)
[error] 	at scala.collection.immutable.HashSet$HashTrieSet.foreach(HashSet.scala:1111)
[error] 	at scala.collection.immutable.HashSet$HashTrieSet.foreach(HashSet.scala:1111)
[error] 	at org.scalajs.linker.frontend.modulesplitter.Tagger.$anonfun$tagEntryPoints$2(Tagger.scala:118)
[error] 	at org.scalajs.linker.frontend.modulesplitter.Tagger.$anonfun$tagEntryPoints$2$adapted(Tagger.scala:117)
[error] 	at scala.collection.TraversableLike$WithFilter.$anonfun$foreach$1(TraversableLike.scala:985)
[error] 	at scala.collection.immutable.Map$Map4.foreach(Map.scala:493)
[error] 	at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:984)
[error] 	at org.scalajs.linker.frontend.modulesplitter.Tagger.tagEntryPoints(Tagger.scala:117)
[error] 	at org.scalajs.linker.frontend.modulesplitter.Tagger.tagAll(Tagger.scala:78)
[error] 	at org.scalajs.linker.frontend.modulesplitter.FewestModulesAnalyzer.analyze(FewestModulesAnalyzer.scala:38)
[error] 	at org.scalajs.linker.frontend.modulesplitter.ModuleSplitter.$anonfun$split$3(ModuleSplitter.scala:51)
[error] 	at org.scalajs.logging.Logger.time(Logger.scala:42)
[error] 	at org.scalajs.logging.Logger.time$(Logger.scala:40)
[error] 	at org.scalajs.sbtplugin.Loggers$SbtLoggerWrapper.time(Loggers.scala:19)
[error] 	at org.scalajs.linker.frontend.modulesplitter.ModuleSplitter.split(ModuleSplitter.scala:51)
[error] 	at org.scalajs.linker.frontend.LinkerFrontendImpl.$anonfun$link$8(LinkerFrontendImpl.scala:77)
[error] 	at org.scalajs.logging.Logger.time(Logger.scala:42)
[error] 	at org.scalajs.logging.Logger.time$(Logger.scala:40)
[error] 	at org.scalajs.sbtplugin.Loggers$SbtLoggerWrapper.time(Loggers.scala:19)
[error] 	at org.scalajs.linker.frontend.LinkerFrontendImpl.$anonfun$link$7(LinkerFrontendImpl.scala:77)
[error] 	at scala.util.Success.$anonfun$map$1(Try.scala:255)
[error] 	at scala.util.Success.map(Try.scala:213)
[error] 	at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
[error] 	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
[error] 	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
[error] 	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
[error] 	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
[error] 	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
[error] 	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
[error] 	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
[error] 	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175)
[error] (testSuite2_12 / Test / fastLinkJS) java.util.NoSuchElementException: key not found: ClassName<org.scalajs.testsuite.jsinterop.ExportedDefaultArgClass$>
```

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala:133]: The tests fail with module splitting, because this line is wrong: The static dependencies of an optimized method are not necessarily the same as the one of an un-optimized method.

In fact: This surfaces a latent issue we've had since module splitting was introduced: The optimizer does not return correct LinkedClasses anymore: The optimizer can (and does) change the real dependencies.

This issue has not been visible before, because the Refiner always recalculated all linking related infos.

Comment (gzm0) [reply linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala:133]: Updated. PTAL. I'm toying with a refactor that might allow the optimizer to not have to lie.

Comment (gzm0) [reply linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala:133]: (for a separate PR that is, IMO this can go in like this nevertheless).

**Review by sjrd — APPROVED** [review]
Not pretty, but we can live with it for now.

**Pull request merged**
